### ISSUE TITLE
`sparse.mul`: broadcasting a size-1 dimension of the sparse operand silently drops data

### ISSUE THREAD

**Issue by mtsokol** [pytorch/pytorch#188900]
### 🐛 Describe the bug

When two tensors are multiplied elementwise and the operand whose size-1 dimension must be broadcast up is a sparse COO tensor, torch.mul silently returns an incorrect result: only the index == 0 slice along the broadcast dimension is populated, and all other slices are zero. No error or warning is raised.

Broadcasting works correctly in every other direction (dense operand broadcast up against a full sparse tensor, equal shapes, scalar multiply). The bug is specific to expanding a singleton dimension of the sparse operand.

## Minimal reproducer

```py
import torch

s  = torch.tensor([[1., 2.], [3., 4.]]).to_sparse_coo()   # (2, 2)
s1 = s[:, :, None]                                        # (2, 2, 1) sparse, trailing singleton
d  = torch.ones(1, 1, 3)                                  # dense, broadcasts result to (2, 2, 3)

out      = (s1 * d).to_dense()
expected = s.to_dense()[:, :, None] * d

print(out)
print(expected)
print("max abs err:", (out - expected).abs().max().item())
```

## Output

```py
max abs err: 4.0
```
`out[..., 0]` matches `expected[..., 0]`, but `out[..., 1:]` is entirely zero, whereas it should equal `out[..., 0]` (broadcasting a size-1 dim replicates the values). The result is silently wrong.

## Expected behavior

Either:
- broadcast the sparse operand's singleton dimension correctly (replicate each nnz across the expanded dimension), matching dense semantics; or
- raise an explicit error if this broadcast pattern is unsupported for the sparse layout — rather than returning a corrupted tensor.

## Real-world impact
Hit while writing an MTTKRP kernel of the form

```py
torch.sum(B[:, :, :, None] * D[None, None, :, :] * C[None, :, None, :], axis=(1, 2))
```
with a sparse COO B. The B[:, :, :, None] term introduces a singleton rank dimension that broadcasts against the dense factor matrices, so the result was silently wrong for every rank column except the first — with no indication anything had gone wrong.

### Versions

PyTorch: 2.12.1+cu130
Python: 3.10.14
CUDA available: yes (bug reproduces on CPU tensors as shown)

cc @ezyang @gchanan @kadeng @msaroufim @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip

**Comment by leninathikam** [comment]
Thanks for the clear reproducer — this is a silent correctness bug.

I traced this to `intersection_binary_op_sparse_dense_out` in `SparseTensorMath.cpp`. The sparse-dense `mul` path handles dense broadcasting over a lower-rank sparse tensor, but it does not explicitly broadcast size-1 dimensions on the sparse operand when ranks already align. In that case it indexes the dense tensor at the sparse indices as-is, so only the slice at index 0 along the broadcast dimension gets values and the rest stay zero.

This looks like the same root cause as #158861, where `@nikitaved` noted that explicit broadcasting via `sparse_broadcast_to` followed by coalesce is missing.

Proposed fix: before index intersection, broadcast the sparse operand to `infer_size(dense, sparse)` using `at::sparse_broadcast_to`, then proceed with the existing logic. I’m planning to put together a PR with regression tests for singleton sparse/dense dims (including an MTTKRP-style case).

Happy to coordinate if someone is already working on this.
